### ISSUE TITLE
[BUG] BackgroundServer Timesout and fails to start

### ISSUE THREAD

**Issue by benallard** [jobrunr/jobrunr#250]
**Are you running the latest version of JobRunr?**
2.0.0

**Describe the bug**
```
2021-10-31 02:12:32.722 [backgroundjob-zookeeper-pool-10-thread-1] INFO  o.jobrunr.server.ServerZooKeeper#deleteServersThatTimedOut[91] - Removed 1 server(s) that timed out
2021-10-31 02:12:32.722 [backgroundjob-zookeeper-pool-10-thread-1] INFO  o.j.server.BackgroundJobServer#setIsMaster[123] - JobRunr BackgroundJobServer (93bc3289-ae3b-465d-8577-90a4cfe0613c) and 5 BackgroundJobPerformers started successfully
2021-10-31 02:12:32.723 [backgroundjob-zookeeper-pool-10-thread-1] INFO  o.jobrunr.server.ServerZooKeeper#determineIfCurrentBackgroundJobServerIsMaster[103] - Server 93bc3289-ae3b-465d-8577-90a4cfe0613c is master (this BackgroundJobServer)
2021-10-31 02:00:32.721 [backgroundjob-zookeeper-pool-10-thread-1] INFO  o.jobrunr.server.ServerZooKeeper#deleteServersThatTimedOut[91] - Removed 1 server(s) that timed out
2021-10-31 02:00:32.722 [backgroundjob-zookeeper-pool-10-thread-1] ERROR o.jobrunr.server.ServerZooKeeper#run[45] - An unrecoverable error occurred. Shutting server down...
org.jobrunr.JobRunrException: JobRunr encountered a problematic exception. Please create a bug report (if possible, provide the code to reproduce this and the stacktrace)
	at org.jobrunr.JobRunrException.shouldNotHappenException(JobRunrException.java:33) ~[jobrunr-2.0.0.jar:na]
	at org.jobrunr.storage.sql.common.BackgroundJobServerTable.lambda$getLongestRunningBackgroundJobServerId$3(BackgroundJobServerTable.java:110) ~[jobrunr-2.0.0.jar:na]
	at java.util.Optional.orElseThrow(Optional.java:290) ~[na:1.8.0_131]
	at org.jobrunr.storage.sql.common.BackgroundJobServerTable.getLongestRunningBackgroundJobServerId(BackgroundJobServerTable.java:110) ~[jobrunr-2.0.0.jar:na]
	at org.jobrunr.storage.sql.common.DefaultSqlStorageProvider.getLongestRunningBackgroundJobServerId(DefaultSqlStorageProvider.java:92) ~[jobrunr-2.0.0.jar:na]
	at org.jobrunr.storage.ThreadSafeStorageProvider.getLongestRunningBackgroundJobServerId(ThreadSafeStorageProvider.java:63) ~[jobrunr-2.0.0.jar:na]
	at org.jobrunr.server.ServerZooKeeper.determineIfCurrentBackgroundJobServerIsMaster(ServerZooKeeper.java:98) ~[jobrunr-2.0.0.jar:na]
	at org.jobrunr.server.ServerZooKeeper.signalBackgroundJobServerAliveAndDoZooKeeping(ServerZooKeeper.java:73) ~[jobrunr-2.0.0.jar:na]
	at org.jobrunr.server.ServerZooKeeper.run(ServerZooKeeper.java:42) ~[jobrunr-2.0.0.jar:na]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_131]
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308) [na:1.8.0_131]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180) [na:1.8.0_131]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294) [na:1.8.0_131]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_131]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_131]
	at java.lang.Thread.run(Thread.java:748) [na:1.8.0_131]
Caused by: java.lang.IllegalStateException: No servers available?!
	... 16 common frames omitted
2021-10-31 02:00:32.723 [backgroundjob-zookeeper-pool-10-thread-1] ERROR o.j.server.BackgroundJobServer#setIsMaster[125] - JobRunr BackgroundJobServer failed to start
```

**Environment**
I'm using JobRunr version: 2.0.0
I'm running on JRE / JDK (e.g. OpenJDK 1.8.0_292): 8
I'm using the following StorageProvider (e.g. Oracle / MongoDB / ...): postgres

**How to Reproduce**
Start server, failed jobs get enqueud, immediately worked on, and then the server claims it failed to start.

**Expected behavior**
It did started, it just failed to recognize it was successfully running.

**Additional context**
The enqueuing of failed jobs upon server startup might be related to the problem.

**Comment by rdehuyss** [comment]
Can you please update to the latest version and confirm whether it is still an issue?

**Comment by benallard** [comment]
I can update it, and we'll see upon the next server rollout. I'm afraid (actually not really) that there won't be any failed jobs to requeue (it's a good sign), so we won't be able to reproduce the exact same situation.

**Comment by benallard** [comment]
As Info, this patch was necessary on our side for the update:

``` patch
===================================================================
--- a/path/to/JobRunRConfiguration.java	(revision 2c2d7bd64f346b0b10367fcdab581f195b065e36)
+++ b/path/to/JobRunRConfiguration.java	(revision e704c48a9038b21cd033c605352d9364c8c00998)
@@ -3,10 +3,13 @@
 import javax.sql.DataSource;
 
 import org.jobrunr.configuration.JobRunr;
+import org.jobrunr.configuration.JobRunrConfiguration;
+import org.jobrunr.scheduling.JobRequestScheduler;
 import org.jobrunr.scheduling.JobScheduler;
 import org.jobrunr.server.BackgroundJobServerConfiguration;
 import org.jobrunr.storage.StorageProvider;
-import org.jobrunr.storage.sql.common.DefaultSqlStorageProvider;
+import org.jobrunr.storage.StorageProviderUtils;
+import org.jobrunr.storage.sql.common.SqlStorageProviderFactory;
 import org.springframework.beans.factory.annotation.Value;
 import org.springframework.context.ApplicationContext;
 import org.springframework.context.annotation.Bean;
@@ -25,13 +28,13 @@
     private int workerCount;
 
     @Bean
-    public JobScheduler initJobRunr(
+    public JobRunrConfiguration.JobRunrConfigurationResult initJobRunr(
             ApplicationContext applicationContext,
             StorageProvider storageProvider) {
         return JobRunr.configure()
                 .useStorageProvider(storageProvider)
                 .useJobActivator(applicationContext::getBean)
-                .useDefaultBackgroundJobServer(BackgroundJobServerConfiguration
+                .useBackgroundJobServer(BackgroundJobServerConfiguration
                         .usingStandardBackgroundJobServerConfiguration()
                         .andWorkerCount(workerCount)
                         .andPollIntervalInSeconds(pollInterval))
@@ -40,8 +43,21 @@
                 .initialize();
     }
 
+    @Bean
+    public JobScheduler initJobScheduler(JobRunrConfiguration.JobRunrConfigurationResult jobRunrConfigurationResult) {
+        return jobRunrConfigurationResult.getJobScheduler();
+    }
+
+    @Bean
+    public JobRequestScheduler initJobRequestScheduler(JobRunrConfiguration.JobRunrConfigurationResult jobRunrConfigurationResult) {
+        return jobRunrConfigurationResult.getJobRequestScheduler();
+    }
+
     @Bean
     public StorageProvider storageProvider(ApplicationContext applicationContext) {
-        return new DefaultSqlStorageProvider(applicationContext.getBean(DataSource.class), DefaultSqlStorageProvider.DatabaseOptions.SKIP_CREATE);
+        return SqlStorageProviderFactory.using(
+                applicationContext.getBean(DataSource.class),
+                null,
+                StorageProviderUtils.DatabaseOptions.SKIP_CREATE);
     }
 }
===================================================================
--- a/path/to/JobRunrRequeueFailedJobs.java	(revision 2c2d7bd64f346b0b10367fcdab581f195b065e36)
+++ b/path/to/JobRunrRequeueFailedJobs.java	(revision e704c48a9038b21cd033c605352d9364c8c00998)
@@ -26,7 +26,7 @@
     @EventListener({ ContextRefreshedEvent.class })
     public void run() {
         log.info("Checking for failed Jobs to re-queue");
-        long total = getJobRunRStorageProvider().countJobs(StateName.FAILED);
+        long total = getJobRunRStorageProvider().getJobStats().getFailed();
         if (total == 0) {
             log.info("Nothing to do.");
             return;

```

**Comment by rdehuyss** [comment]
Since you are using Spring, I would urge you to use the Spring Starter. It will probably result in less breaking changes.

**Comment by benallard** [comment]
I'm using spring, but not spring-boot, so that's an option. (correct me if I'm wrong).

Patch above is missing a DB migration.

**Comment by rdehuyss** [comment]
I just remembered this is now solved since 4.0.

See #192.
